<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>The central data structure of the MultiAmplicon package — MultiAmplicon-class • MultiAmplicon</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="The central data structure of the MultiAmplicon package — MultiAmplicon-class" />
<meta property="og:description" content="The MultiAmplicon class is a container that stores at least primer
pairs, read files and progressively processed data in an 'amplicon
x samples' format. The slots in this object are incrementally
filled with by running wrappers functions (mostly around functions
from the dada2 package). The object is treated (subsetted
etc.) like a (pseudo) matrix, colums are samples, rows are
different amplicons.
Subsetting for MultiAmplicon objects should conveniently subset
all (potentially) filled slots" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MultiAmplicon</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/derele/MultiAmplicon/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>The central data structure of the MultiAmplicon package</h1>
    <small class="dont-index">Source: <a href='https://github.com/derele/MultiAmplicon/blob/master/R/allClasses.R'><code>R/allClasses.R</code></a>, <a href='https://github.com/derele/MultiAmplicon/blob/master/R/show-methods.R'><code>R/show-methods.R</code></a>, <a href='https://github.com/derele/MultiAmplicon/blob/master/R/subset-methods.R'><code>R/subset-methods.R</code></a></small>
    <div class="hidden name"><code>MultiAmplicon-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The MultiAmplicon class is a container that stores at least primer
pairs, read files and progressively processed data in an 'amplicon
x samples' format. The slots in this object are incrementally
filled with by running wrappers functions (mostly around functions
from the <code>dada2</code> package). The object is treated (subsetted
etc.) like a (pseudo) matrix, colums are samples, rows are
different amplicons.</p>
<p>Subsetting for MultiAmplicon objects should conveniently subset
all (potentially) filled slots</p>
    </div>

    <pre class="usage">MultiAmplicon(
  PrimerPairsSet = PrimerPairsSet(),
  PairedReadFileSet = PairedReadFileSet(),
  .Data = matrix(ncol = 0, nrow = 0),
  stratifiedFiles = list(),
  sampleData = new("sample_data", data.frame(row.names = names(PairedReadFileSet),
    readsF = PairedReadFileSet@readsF, readsR = PairedReadFileSet@readsF)),
  derep = list(),
  dada = list(),
  mergers = list(),
  sequenceTable = list(),
  sequenceTableNoChime = list(),
  taxonTable = list()
)

getPrimerPairsSet(MA)

getPairedReadFileSet(MA)

getRawCounts(MA)

getStratifiedFilesF(MA, simplify = TRUE)

getStratifiedFilesR(MA, simplify = TRUE)

getDerepF(MA, simplify = TRUE)

getDerepR(MA, simplify = TRUE)

getDadaF(MA, simplify = TRUE)

getDadaR(MA, simplify = TRUE)

getMergers(MA, simplify = TRUE)

getSequenceTable(MA, simplify = TRUE)

getSequenceTableNoChime(MA, simplify = TRUE)

getTaxonTable(MA, simplify = TRUE)

getSequencesFromTable(MA)

# S4 method for MultiAmplicon
getSequencesFromTable(MA)

# S4 method for MultiAmplicon
show(object)

# S4 method for MultiAmplicon,index,index,ANY
[(x, i, j, ..., drop = FALSE)

# S4 method for MultiAmplicon,index,missing,ANY
[(x, i, j, ..., drop = FALSE)

# S4 method for MultiAmplicon,missing,index,ANY
[(x, i, j, ..., drop = FALSE)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>PrimerPairsSet</th>
      <td><p>a set of primer pairs specifiying your
amplicons see <code><a href='PrimerPairsSet-class.html'>PrimerPairsSet-class</a></code></p></td>
    </tr>
    <tr>
      <th>PairedReadFileSet</th>
      <td><p>a set of paired end sequencing data files
<code><a href='PairedReadFileSet-class.html'>PairedReadFileSet-class</a></code></p></td>
    </tr>
    <tr>
      <th>.Data</th>
      <td><p>Users should not supply this parameter, the slot
is created by <code><a href='sortAmplicons.html'>sortAmplicons</a></code>.</p></td>
    </tr>
    <tr>
      <th>stratifiedFiles</th>
      <td><p>Users should not supply this parameter, the
slot is created by <code><a href='sortAmplicons.html'>sortAmplicons</a></code>.</p></td>
    </tr>
    <tr>
      <th>sampleData</th>
      <td><p>Users should not supply this parameter. It's
filled with a sample_data object from
<code><a href='https://rdrr.io/pkg/phyloseq/man/phyloseq.html'>phyloseq</a></code>. The slot is created from
sample names (same as <code><a href='https://rdrr.io/r/base/colnames.html'>colnames(MA)</a></code>) and more data can
be added by <code><a href='addSampleData.html'>addSampleData</a></code>.</p></td>
    </tr>
    <tr>
      <th>derep</th>
      <td><p>Users should not supply this parameter, the slot is
created by <code><a href='derepMulti.html'>derepMulti</a></code></p></td>
    </tr>
    <tr>
      <th>dada</th>
      <td><p>Users should not supply this parameter, the slot is
created by <code><a href='dadaMulti.html'>dadaMulti</a></code></p></td>
    </tr>
    <tr>
      <th>mergers</th>
      <td><p>Users should not supply this parameter, the slot is
created by <code><a href='mergeMulti.html'>mergeMulti</a></code></p></td>
    </tr>
    <tr>
      <th>sequenceTable</th>
      <td><p>Users should not supply this parameter, the
slot is created by <code><a href='makeSequenceTableMulti.html'>makeSequenceTableMulti</a></code></p></td>
    </tr>
    <tr>
      <th>sequenceTableNoChime</th>
      <td><p>Users should not supply this parameter,
the slot is created by <code><a href='removeChimeraMulti.html'>removeChimeraMulti</a></code></p></td>
    </tr>
    <tr>
      <th>taxonTable</th>
      <td><p>Users should not supply this parameter, the slot
is created by <code><a href='blastTaxAnnot.html'>blastTaxAnnot</a></code>. It's filled with a
list of taxonomyTable objects from
<code><a href='https://rdrr.io/pkg/phyloseq/man/phyloseq.html'>phyloseq</a></code>.</p></td>
    </tr>
    <tr>
      <th>MA</th>
      <td><p>MultiAmplicon-class object</p></td>
    </tr>
    <tr>
      <th>simplify</th>
      <td><p>Should a list of objects be simplified to only one
object if it has length one?</p></td>
    </tr>
    <tr>
      <th>object</th>
      <td><p>A <code>MultiAmplicon-class</code> object.</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>MultiAmplicon-class object</p></td>
    </tr>
    <tr>
      <th>i</th>
      <td><p>numeric, logical or names vector for subsetting rows (==
amplicons)</p></td>
    </tr>
    <tr>
      <th>j</th>
      <td><p>numeric, logical or names vector for subsetting columns
(== read files, corresponding usually to samples)</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>not used</p></td>
    </tr>
    <tr>
      <th>drop</th>
      <td><p>should not be used</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="functions"><a class="anchor" href="#functions"></a>Functions</h2>

    
<ul>
<li><p><code>MultiAmplicon</code>: Constructor for
MultiAmplicon-class</p></li>
</ul>
    <h2 class="hasAnchor" id="slots"><a class="anchor" href="#slots"></a>Slots</h2>

    

<dl>
<dt><code>PrimerPairsSet</code></dt><dd><p>The primer pairs used in your experiment to
specify amplicons stored in a
<code><a href='PrimerPairsSet-class.html'>PrimerPairsSet-class</a></code> object.</p></dd>

<dt><code>PairedReadFileSet</code></dt><dd><p>The (quality filtered) fastq files (one
file pair for each sample) that store your sequencing data.</p></dd>

<dt><code>.Data</code></dt><dd><p>A numeric matrix of sequencing read counts per
amplicon and sample. Created by the function
<code><a href='sortAmplicons.html'>sortAmplicons</a></code> in the MultiAmplicon pipeline.</p></dd>

<dt><code>sampleData</code></dt><dd><p>A sample_data object from
<code><a href='https://rdrr.io/pkg/phyloseq/man/phyloseq.html'>phyloseq</a></code>. The slot is created from
sample names (names of the <code><a href='PrimerPairsSet-class.html'>PrimerPairsSet</a></code>, which
have tto be the same as <code><a href='https://rdrr.io/r/base/colnames.html'>colnames(MA)</a></code>). More data can be
added by <code><a href='addSampleData.html'>addSampleData</a></code>.</p></dd>

<dt><code>stratifiedFiles</code></dt><dd><p>temporary files as a result of stratifying
into amplicons and samples using the MultiAmplicon pipeline
function <code><a href='sortAmplicons.html'>sortAmplicons</a></code>. Forward (sometimes
called R1) and reverse (sometimes called R2) files are stored
as a (amplicons x samples) matrix of
<code><a href='PairedReadFileSet-class.html'>PairedReadFileSet-class</a></code> objects.</p></dd>

<dt><code>derep</code></dt><dd><p>A list of <code><a href='PairedDerep-class.html'>PairedDerep-class</a></code> objects
containing pairs of derep-class objects created by
<code>dada2</code>’s <code><a href='https://rdrr.io/pkg/dada2/man/derepFastq.html'>derepFastq</a></code> function or
withing the MultiAmplicon pipeline by
<code><a href='derepMulti.html'>derepMulti</a></code>.</p></dd>

<dt><code>dada</code></dt><dd><p>A list of <code><a href='PairedDada-class.html'>PairedDada-class</a></code> object
containing pairs of dada-class objects created by
<code>dada2</code>’s <code><a href='https://rdrr.io/pkg/dada2/man/dada.html'>dada</a></code> function. Within the
MultiAmplicon pipeline this slot is filled by
<code><a href='dadaMulti.html'>dadaMulti</a></code>.</p></dd>

<dt><code>mergers</code></dt><dd><p>A list of objects containing merged pairs of forward
and reverse reads as created by by <code>dada2</code>’s
<code><a href='https://rdrr.io/pkg/dada2/man/mergePairs.html'>mergePairs</a></code> function. Within the
MultiAmplicon pipeline this slot is filled by
<code><a href='mergeMulti.html'>mergeMulti</a></code>.</p></dd>

<dt><code>sequenceTable</code></dt><dd><p>A list of matrix objects created by
<code>dada2</code>’s <code><a href='https://rdrr.io/pkg/dada2/man/makeSequenceTable.html'>makeSequenceTable</a></code>.
Samples (in rows) and amplified sequence variants (ASVs) in
columns.  Within the MultiAmplicon pipeline this slot is
filled by <code><a href='makeSequenceTableMulti.html'>makeSequenceTableMulti</a></code>.</p></dd>

<dt><code>sequenceTableNoChime</code></dt><dd><p>A list of matrix objects created by
<code>dada2</code>’s <code><a href='https://rdrr.io/pkg/dada2/man/removeBimeraDenovo.html'>removeBimeraDenovo</a></code>.
Samples (in rows) and ASVs screened for PCR chimeras in
columns. Within the MultiAmplicon pipeline this slot is filled
by <code><a href='removeChimeraMulti.html'>removeChimeraMulti</a></code>.</p></dd>

<dt><code>taxonTable</code></dt><dd><p>A list of matrix objects created by a function
    for taxonomical annotation (for example
    <code><a href='blastTaxAnnot.html'>blastTaxAnnot</a></code>.  ASVs are in rows and taxnomical
    ranks are in columns. MultiAmplicon(PrimerPairsSet, PairedReadFileSet)</p></dd>

</dl>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='https://rdrr.io/pkg/dada2/man/derepFastq.html'>derepFastq</a></code>,<code><a href='https://rdrr.io/pkg/dada2/man/dada.html'>dada</a></code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Emanuel Heitlinger</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='va'>primerF</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"AGAGTTTGATCCTGGCTCAG"</span>, <span class='st'>"ACTCCTACGGGAGGCAGC"</span>,
            <span class='st'>"GAATTGACGGAAGGGCACC"</span>, <span class='st'>"YGGTGRTGCATGGCCGYT"</span><span class='op'>)</span>
<span class='va'>primerR</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"CTGCWGCCNCCCGTAGG"</span>, <span class='st'>"GACTACHVGGGTATCTAATCC"</span>,
             <span class='st'>"AAGGGCATCACAGACCTGTTAT"</span>, <span class='st'>"TCCTTCTGCAGGTTCACCTAC"</span><span class='op'>)</span>

<span class='va'>PPS</span> <span class='op'>&lt;-</span> <span class='fu'><a href='PrimerPairsSet-class.html'>PrimerPairsSet</a></span><span class='op'>(</span><span class='va'>primerF</span>, <span class='va'>primerR</span><span class='op'>)</span>

<span class='va'>fastq.dir</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"fastq"</span>, package <span class='op'>=</span> <span class='st'>"MultiAmplicon"</span><span class='op'>)</span>
<span class='va'>fastq.files</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.files.html'>list.files</a></span><span class='op'>(</span><span class='va'>fastq.dir</span>, full.names<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>
<span class='va'>Ffastq.file</span> <span class='op'>&lt;-</span> <span class='va'>fastq.files</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/grep.html'>grepl</a></span><span class='op'>(</span><span class='st'>"F_filt"</span>, <span class='va'>fastq.files</span><span class='op'>)</span><span class='op'>]</span>
<span class='va'>Rfastq.file</span> <span class='op'>&lt;-</span> <span class='va'>fastq.files</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/grep.html'>grepl</a></span><span class='op'>(</span><span class='st'>"R_filt"</span>, <span class='va'>fastq.files</span><span class='op'>)</span><span class='op'>]</span>

<span class='va'>PRF</span> <span class='op'>&lt;-</span> <span class='fu'><a href='PairedReadFileSet-class.html'>PairedReadFileSet</a></span><span class='op'>(</span><span class='va'>Ffastq.file</span>, <span class='va'>Rfastq.file</span><span class='op'>)</span>

<span class='va'>MA</span> <span class='op'>&lt;-</span> <span class='fu'>MultiAmplicon</span><span class='op'>(</span><span class='va'>PPS</span>, <span class='va'>PRF</span><span class='op'>)</span>

<span class='co'>## sort into amplicons</span>
<span class='va'>MA1</span> <span class='op'>&lt;-</span> <span class='fu'><a href='sortAmplicons.html'>sortAmplicons</a></span><span class='op'>(</span><span class='va'>MA</span>, filedir<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempfile</a></span><span class='op'>(</span>pattern <span class='op'>=</span> <span class='st'>"dir"</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>creating directory /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T//RtmpmUDm5W/diracf797a300a</span></div><div class='output co'>#&gt; <span class='message'>finished sorting 0 sequencing reads for S00_F_filt.fastq.gz in </span>
#&gt; <span class='message'> /Users/runner/work/_temp/Library/MultiAmplicon/extdata/fastq/S00_F_filt.fastq.gz and </span>
#&gt; <span class='message'>  /Users/runner/work/_temp/Library/MultiAmplicon/extdata/fastq/S00_R_filt.fastq.gz </span>
#&gt; <span class='message'> into 0  amplicons and written into /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpmUDm5W/diracf797a300a</span></div><div class='output co'>#&gt; <span class='message'>finished sorting 0 sequencing reads for S01_F_filt.fastq.gz in </span>
#&gt; <span class='message'> /Users/runner/work/_temp/Library/MultiAmplicon/extdata/fastq/S01_F_filt.fastq.gz and </span>
#&gt; <span class='message'>  /Users/runner/work/_temp/Library/MultiAmplicon/extdata/fastq/S01_R_filt.fastq.gz </span>
#&gt; <span class='message'> into 0  amplicons and written into /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpmUDm5W/diracf797a300a</span></div><div class='output co'>#&gt; <span class='message'>finished sorting 9348 sequencing reads for S02_F_filt.fastq.gz in </span>
#&gt; <span class='message'> /Users/runner/work/_temp/Library/MultiAmplicon/extdata/fastq/S02_F_filt.fastq.gz and </span>
#&gt; <span class='message'>  /Users/runner/work/_temp/Library/MultiAmplicon/extdata/fastq/S02_R_filt.fastq.gz </span>
#&gt; <span class='message'> into 4  amplicons and written into /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpmUDm5W/diracf797a300a</span></div><div class='output co'>#&gt; <span class='message'>finished sorting 5 sequencing reads for S03_F_filt.fastq.gz in </span>
#&gt; <span class='message'> /Users/runner/work/_temp/Library/MultiAmplicon/extdata/fastq/S03_F_filt.fastq.gz and </span>
#&gt; <span class='message'>  /Users/runner/work/_temp/Library/MultiAmplicon/extdata/fastq/S03_R_filt.fastq.gz </span>
#&gt; <span class='message'> into 1  amplicons and written into /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpmUDm5W/diracf797a300a</span></div><div class='output co'>#&gt; <span class='message'>finished sorting 14 sequencing reads for S04_F_filt.fastq.gz in </span>
#&gt; <span class='message'> /Users/runner/work/_temp/Library/MultiAmplicon/extdata/fastq/S04_F_filt.fastq.gz and </span>
#&gt; <span class='message'>  /Users/runner/work/_temp/Library/MultiAmplicon/extdata/fastq/S04_R_filt.fastq.gz </span>
#&gt; <span class='message'> into 3  amplicons and written into /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpmUDm5W/diracf797a300a</span></div><div class='output co'>#&gt; <span class='message'>finished sorting 51777 sequencing reads for S05_F_filt.fastq.gz in </span>
#&gt; <span class='message'> /Users/runner/work/_temp/Library/MultiAmplicon/extdata/fastq/S05_F_filt.fastq.gz and </span>
#&gt; <span class='message'>  /Users/runner/work/_temp/Library/MultiAmplicon/extdata/fastq/S05_R_filt.fastq.gz </span>
#&gt; <span class='message'> into 3  amplicons and written into /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpmUDm5W/diracf797a300a</span></div><div class='output co'>#&gt; <span class='message'>finished sorting 51777 sequencing reads for S05D_F_filt.fastq.gz in </span>
#&gt; <span class='message'> /Users/runner/work/_temp/Library/MultiAmplicon/extdata/fastq/S05D_F_filt.fastq.gz and </span>
#&gt; <span class='message'>  /Users/runner/work/_temp/Library/MultiAmplicon/extdata/fastq/S05D_R_filt.fastq.gz </span>
#&gt; <span class='message'> into 3  amplicons and written into /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpmUDm5W/diracf797a300a</span></div><div class='output co'>#&gt; <span class='message'>finished sorting 17358 sequencing reads for S06_F_filt.fastq.gz in </span>
#&gt; <span class='message'> /Users/runner/work/_temp/Library/MultiAmplicon/extdata/fastq/S06_F_filt.fastq.gz and </span>
#&gt; <span class='message'>  /Users/runner/work/_temp/Library/MultiAmplicon/extdata/fastq/S06_R_filt.fastq.gz </span>
#&gt; <span class='message'> into 4  amplicons and written into /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/RtmpmUDm5W/diracf797a300a</span></div><div class='input'>
<span class='co'>## Only after sorting the MultiAmplicon object is really poplated</span>
<span class='co'>## with sensible data, now matrix-like access to different </span>
<span class='co'>## amplicons (primer pairs) and different sequencing read files</span>
<span class='co'>## (usually samples) is implemented</span>

<span class='co'>## the number of amplicons (primer pairs)</span>
<span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>MA</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1] 0</div><div class='input'>
<span class='co'>## the number of samples (sequencing read file pairs)</span>
<span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>ncol</a></span><span class='op'>(</span><span class='va'>MA</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1] 0</div><div class='input'>
<span class='co'>## dereplication is currently not supported</span>
<span class='co'>## MA2 &lt;- derepMulti(MA1)</span>

<span class='co'>### use dada directly after sorting</span>
<span class='va'>MA3</span> <span class='op'>&lt;-</span> <span class='fu'><a href='dadaMulti.html'>dadaMulti</a></span><span class='op'>(</span><span class='va'>MA1</span>, selfConsist <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>amplicon AGAGTTTGATCCTGGCTCAG.CTGCWGCCNCCCGTAGG: dada estimation of sequence variants from 6 of 8 possible sample files</span></div><div class='output co'>#&gt; <span class='message'>calling dada with selfConsist=TRUE parameters</span></div><div class='output co'>#&gt; Initializing error rates to maximum possible estimate.
#&gt; selfConsist step 1 ......
#&gt;    selfConsist step 2
#&gt;    selfConsist step 3
#&gt;    selfConsist step 4
#&gt;    selfConsist step 5
#&gt;    selfConsist step 6
#&gt; Convergence after  6  rounds.
#&gt; Initializing error rates to maximum possible estimate.
#&gt; selfConsist step 1 ......
#&gt;    selfConsist step 2
#&gt;    selfConsist step 3
#&gt;    selfConsist step 4
#&gt;    selfConsist step 5
#&gt; Convergence after  5  rounds.</div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>amplicon ACTCCTACGGGAGGCAGC.GACTACHVGGGTATCTAATCC: dada estimation of sequence variants from 5 of 8 possible sample files</span></div><div class='output co'>#&gt; <span class='message'>calling dada with selfConsist=TRUE parameters</span></div><div class='output co'>#&gt; Initializing error rates to maximum possible estimate.
#&gt; selfConsist step 1 .....
#&gt;    selfConsist step 2
#&gt;    selfConsist step 3
#&gt;    selfConsist step 4
#&gt;    selfConsist step 5
#&gt;    selfConsist step 6
#&gt; Convergence after  6  rounds.
#&gt; Initializing error rates to maximum possible estimate.
#&gt; selfConsist step 1 .....
#&gt;    selfConsist step 2
#&gt;    selfConsist step 3
#&gt;    selfConsist step 4
#&gt;    selfConsist step 5
#&gt; Convergence after  5  rounds.</div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>amplicon GAATTGACGGAAGGGCACC.AAGGGCATCACAGACCTGTTAT: dada estimation of sequence variants from 2 of 8 possible sample files</span></div><div class='output co'>#&gt; <span class='message'>calling dada with selfConsist=TRUE parameters</span></div><div class='output co'>#&gt; Initializing error rates to maximum possible estimate.
#&gt; selfConsist step 1 ..
#&gt;    selfConsist step 2
#&gt;    selfConsist step 3
#&gt;    selfConsist step 4
#&gt;    selfConsist step 5
#&gt; Convergence after  5  rounds.
#&gt; Initializing error rates to maximum possible estimate.
#&gt; selfConsist step 1 ..
#&gt;    selfConsist step 2
#&gt;    selfConsist step 3
#&gt;    selfConsist step 4
#&gt;    selfConsist step 5
#&gt; Convergence after  5  rounds.</div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>amplicon YGGTGRTGCATGGCCGYT.TCCTTCTGCAGGTTCACCTAC: dada estimation of sequence variants from 5 of 8 possible sample files</span></div><div class='output co'>#&gt; <span class='message'>calling dada with selfConsist=TRUE parameters</span></div><div class='output co'>#&gt; Initializing error rates to maximum possible estimate.
#&gt; selfConsist step 1 .....
#&gt;    selfConsist step 2
#&gt;    selfConsist step 3
#&gt;    selfConsist step 4
#&gt;    selfConsist step 5
#&gt; Convergence after  5  rounds.
#&gt; Initializing error rates to maximum possible estimate.
#&gt; selfConsist step 1 .....
#&gt;    selfConsist step 2
#&gt;    selfConsist step 3
#&gt;    selfConsist step 4
#&gt; Convergence after  4  rounds.</div><div class='input'>
<span class='va'>MA4</span> <span class='op'>&lt;-</span> <span class='fu'><a href='mergeMulti.html'>mergeMulti</a></span><span class='op'>(</span><span class='va'>MA3</span>, justConcatenate<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>merging sequences from 6 samples for amplicon AGAGTTTGATCCTGGCTCAG.CTGCWGCCNCCCGTAGG</span></div><div class='output co'>#&gt; <span class='message'>calling mergePairs with justConcatenate=TRUE parameters</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>merging sequences from 5 samples for amplicon ACTCCTACGGGAGGCAGC.GACTACHVGGGTATCTAATCC</span></div><div class='output co'>#&gt; <span class='message'>calling mergePairs with justConcatenate=TRUE parameters</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>merging sequences from 2 samples for amplicon GAATTGACGGAAGGGCACC.AAGGGCATCACAGACCTGTTAT</span></div><div class='output co'>#&gt; <span class='message'>calling mergePairs with justConcatenate=TRUE parameters</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>merging sequences from 5 samples for amplicon YGGTGRTGCATGGCCGYT.TCCTTCTGCAGGTTCACCTAC</span></div><div class='output co'>#&gt; <span class='message'>calling mergePairs with justConcatenate=TRUE parameters</span></div><div class='input'>
<span class='va'>MA5</span> <span class='op'>&lt;-</span> <span class='fu'><a href='makeSequenceTableMulti.html'>makeSequenceTableMulti</a></span><span class='op'>(</span><span class='va'>MA4</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>calling makeSequenceTable with default parameters</span></div><div class='output co'>#&gt; <span class='message'>calling makeSequenceTable with default parameters</span></div><div class='output co'>#&gt; <span class='message'>calling makeSequenceTable with default parameters</span></div><div class='output co'>#&gt; <span class='message'>calling makeSequenceTable with default parameters</span></div><div class='input'>
<span class='va'>MA6</span> <span class='op'>&lt;-</span> <span class='fu'><a href='removeChimeraMulti.html'>removeChimeraMulti</a></span><span class='op'>(</span><span class='va'>MA5</span>, mc.cores<span class='op'>=</span><span class='fl'>1</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>calling removeBimeraDenovo with default parameters</span></div><div class='output co'>#&gt; <span class='message'>calling removeBimeraDenovo with default parameters</span></div><div class='output co'>#&gt; <span class='message'>calling removeBimeraDenovo with default parameters</span></div><div class='output co'>#&gt; <span class='message'>calling removeBimeraDenovo with default parameters</span></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Emanuel Heitlinger.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


