<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>MultiAmplicon - a small example • MultiAmplicon</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="MultiAmplicon - a small example">
<meta property="og:description" content="MultiAmplicon">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MultiAmplicon</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/MultiAmplicon-real-world-example.html">MultiAmplicon - a real world example</a>
    </li>
    <li>
      <a href="../articles/MultiAmplicon-small-example.html">MultiAmplicon - a small example</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/derele/MultiAmplicon/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><link href="MultiAmplicon-small-example_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="MultiAmplicon-small-example_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>MultiAmplicon - a small example</h1>
                        <h4 class="author">Emanuel Heitlinger</h4>
            
            <h4 class="date">2020-11-26</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/derele/MultiAmplicon/blob/master/vignettes/MultiAmplicon-small-example.Rmd"><code>vignettes/MultiAmplicon-small-example.Rmd</code></a></small>
      <div class="hidden name"><code>MultiAmplicon-small-example.Rmd</code></div>

    </div>

    
    
<div id="preparing-sequencing-read-files-and-a-primer-set" class="section level2">
<h2 class="hasAnchor">
<a href="#preparing-sequencing-read-files-and-a-primer-set" class="anchor"></a>Preparing sequencing read files and a primer set</h2>
<p>The MultiAmplicon package uses dedicated classes to ensure that input to its <code>sortAmplicons</code> function is provided correctly. Concerning input files, this means for the user that file names (including paths to the directory they are stored in) of forward and reverse sequencing read files must be provided as a <code>PairedReadFileSet</code>. These files can contain reads in fastq or in gzipped fastq format. We focus as a default use case on (non-interleaved) paired-end reads, as these are the most common format Illumina sequencing reads are currently produced in.</p>
<p>Similarly, primer pairs must be provided in a <code>PrimerPairsSet</code>. We then match forward and reverse primers at the start of forward and reverse reads, respectively. The maximal number of mismatches and the maximal distance from the start of the read can be supplied as arguments to the function. The primer sequences are trimmed from the reads and the read-pair is associated with the amplicon defined by the primer-pair.</p>
<p>We first have a look at how <code>PairedReadFileSet</code> and <code>PrimerPairsSet</code> are generated. Please use files containing quality filtered sequencing reads for the steps below. See the <a href="https://benjjneb.github.io/dada2/tutorial.html">dada2 tutorial</a> or our <a href="https://derele.github.io/MultiAmplicon/articles/MultiAmplicon-real-world-example.html">real-world example vignette</a> for how files are trimmed and quality filtered. But we here first use a very small data set included in the MultiAmplicon package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://derele.github.io/MultiAmplicon/">MultiAmplicon</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: replacing previous import 'data.table:::=' by 'ggplot2:::=' when</span>
<span class="co">#&gt; loading 'MultiAmplicon'</span>

<span class="va">fastq.dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"fastq"</span>, package <span class="op">=</span> <span class="st">"MultiAmplicon"</span><span class="op">)</span>
<span class="va">fastq.files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">fastq.dir</span>, full.names<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>

<span class="va">Ffastq.file</span> <span class="op">&lt;-</span> <span class="va">fastq.files</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"F_filt"</span>, <span class="va">fastq.files</span><span class="op">)</span><span class="op">]</span>
<span class="va">Rfastq.file</span> <span class="op">&lt;-</span> <span class="va">fastq.files</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"R_filt"</span>, <span class="va">fastq.files</span><span class="op">)</span><span class="op">]</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">Rfastq.file</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">Ffastq.file</span><span class="op">)</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"_F_filt\\.fastq\\.gz"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">Ffastq.file</span><span class="op">)</span><span class="op">)</span>

<span class="va">PRF</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PairedReadFileSet-class.html">PairedReadFileSet</a></span><span class="op">(</span><span class="va">Ffastq.file</span>, <span class="va">Rfastq.file</span><span class="op">)</span>

<span class="va">primerF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>Amp1F <span class="op">=</span> <span class="st">"AGAGTTTGATCCTGGCTCAG"</span>, Amp2F <span class="op">=</span> <span class="st">"ACTCCTACGGGAGGCAGC"</span>,
             Amp3F <span class="op">=</span> <span class="st">"GAATTGACGGAAGGGCACC"</span>, Amp4F <span class="op">=</span> <span class="st">"YGGTGRTGCATGGCCGYT"</span>,
             Amp5F <span class="op">=</span> <span class="st">"AAAAACCCCGGGGGGTTTTT"</span>, Amp6F <span class="op">=</span> <span class="st">"AGAGTTTGATCCTGCCTCAG"</span><span class="op">)</span>

<span class="va">primerR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>Amp1R <span class="op">=</span> <span class="st">"CTGCWGCCNCCCGTAGG"</span>, Amp2R <span class="op">=</span> <span class="st">"GACTACHVGGGTATCTAATCC"</span>,
             Amp3R <span class="op">=</span> <span class="st">"AAGGGCATCACAGACCTGTTAT"</span>, Amp4R <span class="op">=</span> <span class="st">"TCCTTCTGCAGGTTCACCTAC"</span>,
             Amp5R <span class="op">=</span> <span class="st">"AAAAACCCCGGGGGGTTTTT"</span>, Amp6R <span class="op">=</span> <span class="st">"CCTACGGGTGGCAGATGCAG"</span><span class="op">)</span>

<span class="va">PPS</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PrimerPairsSet-class.html">PrimerPairsSet</a></span><span class="op">(</span><span class="va">primerF</span>, <span class="va">primerR</span><span class="op">)</span>

<span class="va">MA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MultiAmplicon-class.html">MultiAmplicon</a></span><span class="op">(</span><span class="va">PPS</span>, <span class="va">PRF</span><span class="op">)</span></code></pre></div>
<p>Note that the primer pairs in the <code>PrimerPairsSet</code> can be named and the same also is true for the sequencing read file pairs (although not in our example above). This results in names for primer-pairs and for samples.</p>
</div>
<div id="sorting-into-multiple-amplicons" class="section level2">
<h2 class="hasAnchor">
<a href="#sorting-into-multiple-amplicons" class="anchor"></a>Sorting into multiple amplicons</h2>
<p>Based on the primer sequences we can now sort the sequences into amplicons. This is done by the function <code>sortAmplicons</code>, which updates the MultiAmplicon object with the sorted amplicons.</p>
<p>For this small example data set we can look at the resulting <code>rawCounts</code> (function <code>getRawCounts</code>) i.e. the number of sequences for each of the fully stratified amplicon x sample matrix.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">MA1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sortAmplicons.html">sortAmplicons</a></span><span class="op">(</span><span class="va">MA</span>, filedir<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span><span class="op">(</span><span class="st">"dir"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="fu"><a href="../reference/MultiAmplicon-class.html">getRawCounts</a></span><span class="op">(</span><span class="va">MA1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">S00</th>
<th align="right">S01</th>
<th align="right">S02</th>
<th align="right">S03</th>
<th align="right">S04</th>
<th align="right">S05</th>
<th align="right">S05D</th>
<th align="right">S06</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Amp1F.Amp1R</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">3495</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">29310</td>
<td align="right">29310</td>
<td align="right">2569</td>
</tr>
<tr class="even">
<td align="left">Amp2F.Amp2R</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">2949</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">20391</td>
<td align="right">20391</td>
<td align="right">2690</td>
</tr>
<tr class="odd">
<td align="left">Amp3F.Amp3R</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1817</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">11671</td>
</tr>
<tr class="even">
<td align="left">Amp4F.Amp4R</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1087</td>
<td align="right">0</td>
<td align="right">7</td>
<td align="right">2076</td>
<td align="right">2076</td>
<td align="right">428</td>
</tr>
<tr class="odd">
<td align="left">Amp5F.Amp5R</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">Amp6F.Amp6R</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
</div>
<div id="visualizing-read-numbers-per-amplicon-and-filtering-based-on-these" class="section level2">
<h2 class="hasAnchor">
<a href="#visualizing-read-numbers-per-amplicon-and-filtering-based-on-these" class="anchor"></a>Visualizing read numbers per amplicon and filtering based on these</h2>
<p>Now we can plot the result. The function plotAmpliconNumbers allows us also to store clustering information (see <a href="https://www.rdocumentation.org/packages/pheatmap/versions/0.2/topics/pheatmap"></a>). Which can afterwards be used to filter data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotAmpliconNumbers.html">plotAmpliconNumbers</a></span><span class="op">(</span><span class="va">MA1</span><span class="op">)</span></code></pre></div>
<p><img src="MultiAmplicon-small-example_files/figure-html/plotAmpliconNumbers-1.png" width="700"></p>
<p>We can use this clustering information now to subset our MultiAmplicon object.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">two.clusters.row</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html">cutree</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">$</span><span class="va">tree_row</span>, k<span class="op">=</span><span class="fl">2</span><span class="op">)</span>
<span class="va">two.clusters.col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html">cutree</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">$</span><span class="va">tree_col</span>, k<span class="op">=</span><span class="fl">2</span><span class="op">)</span>

<span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="va">two.clusters.row</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">x</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Amp1F.Amp1R</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">Amp2F.Amp2R</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">Amp3F.Amp3R</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td align="left">Amp4F.Amp4R</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">Amp5F.Amp5R</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td align="left">Amp6F.Amp6R</td>
<td align="right">2</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="va">two.clusters.col</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">x</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">S00</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">S01</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">S02</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td align="left">S03</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">S04</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">S05</td>
<td align="right">2</td>
</tr>
<tr class="odd">
<td align="left">S05D</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td align="left">S06</td>
<td align="right">2</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">MA.sub</span> <span class="op">&lt;-</span> <span class="va">MA1</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">two.clusters.row</span><span class="op">==</span><span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">two.clusters.col</span><span class="op">==</span><span class="fl">2</span><span class="op">)</span><span class="op">]</span>

<span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="fu"><a href="../reference/MultiAmplicon-class.html">getRawCounts</a></span><span class="op">(</span><span class="va">MA.sub</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">S02</th>
<th align="right">S05</th>
<th align="right">S05D</th>
<th align="right">S06</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Amp1F.Amp1R</td>
<td align="right">3495</td>
<td align="right">29310</td>
<td align="right">29310</td>
<td align="right">2569</td>
</tr>
<tr class="even">
<td align="left">Amp2F.Amp2R</td>
<td align="right">2949</td>
<td align="right">20391</td>
<td align="right">20391</td>
<td align="right">2690</td>
</tr>
<tr class="odd">
<td align="left">Amp4F.Amp4R</td>
<td align="right">1087</td>
<td align="right">2076</td>
<td align="right">2076</td>
<td align="right">428</td>
</tr>
</tbody>
</table>
<p>In real world applications (see <a href="https://derele.github.io/MultiAmplicon/articles/MultiAmplicon-real-world-example.html">our other vignette</a>) the samples to be excluded from further analysis can be selected based on clustering with negative controls (similar to the example data here where S00 is an empty file and S01 contains one sequence).</p>
</div>
<div id="running-the-amplicon-sequencing-pipeline" class="section level2">
<h2 class="hasAnchor">
<a href="#running-the-amplicon-sequencing-pipeline" class="anchor"></a>Running the amplicon sequencing pipeline</h2>
<p>We first estimate the errors separately for both for dada’s probabilistic sequence variant inference. We do this here on the combined stratified files for different amplicons. This should be done separately for forward ad reverse sequence though!</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">errF</span> <span class="op">&lt;-</span> <span class="fu">learnErrors</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="../reference/MultiAmplicon-class.html">getStratifiedFilesF</a></span><span class="op">(</span><span class="va">MA1</span><span class="op">)</span><span class="op">)</span>, verbose<span class="op">=</span><span class="fl">0</span><span class="op">)</span>
<span class="va">errR</span> <span class="op">&lt;-</span> <span class="fu">learnErrors</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="../reference/MultiAmplicon-class.html">getStratifiedFilesR</a></span><span class="op">(</span><span class="va">MA1</span><span class="op">)</span><span class="op">)</span>, verbose<span class="op">=</span><span class="fl">0</span><span class="op">)</span></code></pre></div>
<p>We can now run the pipeline of the dada2 package in parallel over the different amplicons. We update our object from each previous step and safe it under a new name, here for illustration. In your own workflow you can also overwrite the old object.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">MA3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dadaMulti.html">dadaMulti</a></span><span class="op">(</span><span class="va">MA1</span>, Ferr<span class="op">=</span><span class="va">errF</span>, Rerr<span class="op">=</span><span class="va">errR</span>,  pool<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt; Sample 1 - 3495 reads in 1493 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 5 reads in 5 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 5 reads in 5 unique sequences.</span>
<span class="co">#&gt; Sample 4 - 29310 reads in 7224 unique sequences.</span>
<span class="co">#&gt; Sample 5 - 29310 reads in 7224 unique sequences.</span>
<span class="co">#&gt; Sample 6 - 2569 reads in 1053 unique sequences.</span>
<span class="co">#&gt; Sample 1 - 3495 reads in 1535 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 5 reads in 5 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 5 reads in 4 unique sequences.</span>
<span class="co">#&gt; Sample 4 - 29310 reads in 8273 unique sequences.</span>
<span class="co">#&gt; Sample 5 - 29310 reads in 8273 unique sequences.</span>
<span class="co">#&gt; Sample 6 - 2569 reads in 994 unique sequences.</span>
<span class="co">#&gt; Sample 1 - 2949 reads in 1142 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 2 reads in 2 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 20391 reads in 4136 unique sequences.</span>
<span class="co">#&gt; Sample 4 - 20391 reads in 4136 unique sequences.</span>
<span class="co">#&gt; Sample 5 - 2690 reads in 968 unique sequences.</span>
<span class="co">#&gt; Sample 1 - 2949 reads in 1290 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 2 reads in 2 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 20391 reads in 5719 unique sequences.</span>
<span class="co">#&gt; Sample 4 - 20391 reads in 5719 unique sequences.</span>
<span class="co">#&gt; Sample 5 - 2690 reads in 1100 unique sequences.</span>
<span class="co">#&gt; Sample 1 - 1817 reads in 589 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 11671 reads in 3783 unique sequences.</span>
<span class="co">#&gt; Sample 1 - 1817 reads in 547 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 11671 reads in 4054 unique sequences.</span>
<span class="co">#&gt; Sample 1 - 1087 reads in 394 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 7 reads in 6 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 2076 reads in 959 unique sequences.</span>
<span class="co">#&gt; Sample 4 - 2076 reads in 959 unique sequences.</span>
<span class="co">#&gt; Sample 5 - 428 reads in 166 unique sequences.</span>
<span class="co">#&gt; Sample 1 - 1087 reads in 433 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 7 reads in 5 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 2076 reads in 962 unique sequences.</span>
<span class="co">#&gt; Sample 4 - 2076 reads in 962 unique sequences.</span>
<span class="co">#&gt; Sample 5 - 428 reads in 183 unique sequences.</span>

<span class="va">MA4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mergeMulti.html">mergeMulti</a></span><span class="op">(</span><span class="va">MA3</span>, justConcatenate<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>

<span class="va">MA5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeSequenceTableMulti.html">makeSequenceTableMulti</a></span><span class="op">(</span><span class="va">MA4</span><span class="op">)</span>

<span class="va">MA6</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/removeChimeraMulti.html">removeChimeraMulti</a></span><span class="op">(</span><span class="va">MA5</span>, mc.cores<span class="op">=</span><span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>The steps after <code>sortAmplcions</code> until <code>removeChimeraMulti</code> recapitulate different processes in the <a href="https://benjjneb.github.io/dada2/tutorial.html">dada2 pipeline</a>.</p>
</div>
<div id="a-word-on-dada2-sequence-inference" class="section level2">
<h2 class="hasAnchor">
<a href="#a-word-on-dada2-sequence-inference" class="anchor"></a>A word on dada2 sequence inference</h2>
<p>The above workflow assumes that errors could be estimated for each amplicon together. This would be beneficial if errors were largely similar between different amplicons (i.e. not influenced by the removed primer having starting the original sequence). Below an alternative approach is demonstrated with error probabilities estimated separately for each amplicon (here also overwrite objects from each previous processing steps).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">MA.alt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dadaMulti.html">dadaMulti</a></span><span class="op">(</span><span class="va">MA3</span>, selfConsist<span class="op">=</span><span class="cn">TRUE</span>, pool<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt; Initializing error rates to maximum possible estimate.</span>
<span class="co">#&gt; selfConsist step 1 ......</span>
<span class="co">#&gt;    selfConsist step 2</span>
<span class="co">#&gt;    selfConsist step 3</span>
<span class="co">#&gt;    selfConsist step 4</span>
<span class="co">#&gt;    selfConsist step 5</span>
<span class="co">#&gt;    selfConsist step 6</span>
<span class="co">#&gt; Convergence after  6  rounds.</span>
<span class="co">#&gt; Initializing error rates to maximum possible estimate.</span>
<span class="co">#&gt; selfConsist step 1 ......</span>
<span class="co">#&gt;    selfConsist step 2</span>
<span class="co">#&gt;    selfConsist step 3</span>
<span class="co">#&gt;    selfConsist step 4</span>
<span class="co">#&gt;    selfConsist step 5</span>
<span class="co">#&gt; Convergence after  5  rounds.</span>
<span class="co">#&gt; Initializing error rates to maximum possible estimate.</span>
<span class="co">#&gt; selfConsist step 1 .....</span>
<span class="co">#&gt;    selfConsist step 2</span>
<span class="co">#&gt;    selfConsist step 3</span>
<span class="co">#&gt;    selfConsist step 4</span>
<span class="co">#&gt;    selfConsist step 5</span>
<span class="co">#&gt;    selfConsist step 6</span>
<span class="co">#&gt; Convergence after  6  rounds.</span>
<span class="co">#&gt; Initializing error rates to maximum possible estimate.</span>
<span class="co">#&gt; selfConsist step 1 .....</span>
<span class="co">#&gt;    selfConsist step 2</span>
<span class="co">#&gt;    selfConsist step 3</span>
<span class="co">#&gt;    selfConsist step 4</span>
<span class="co">#&gt;    selfConsist step 5</span>
<span class="co">#&gt; Convergence after  5  rounds.</span>
<span class="co">#&gt; Initializing error rates to maximum possible estimate.</span>
<span class="co">#&gt; selfConsist step 1 ..</span>
<span class="co">#&gt;    selfConsist step 2</span>
<span class="co">#&gt;    selfConsist step 3</span>
<span class="co">#&gt;    selfConsist step 4</span>
<span class="co">#&gt;    selfConsist step 5</span>
<span class="co">#&gt; Convergence after  5  rounds.</span>
<span class="co">#&gt; Initializing error rates to maximum possible estimate.</span>
<span class="co">#&gt; selfConsist step 1 ..</span>
<span class="co">#&gt;    selfConsist step 2</span>
<span class="co">#&gt;    selfConsist step 3</span>
<span class="co">#&gt;    selfConsist step 4</span>
<span class="co">#&gt;    selfConsist step 5</span>
<span class="co">#&gt; Convergence after  5  rounds.</span>
<span class="co">#&gt; Initializing error rates to maximum possible estimate.</span>
<span class="co">#&gt; selfConsist step 1 .....</span>
<span class="co">#&gt;    selfConsist step 2</span>
<span class="co">#&gt;    selfConsist step 3</span>
<span class="co">#&gt;    selfConsist step 4</span>
<span class="co">#&gt;    selfConsist step 5</span>
<span class="co">#&gt; Convergence after  5  rounds.</span>
<span class="co">#&gt; Initializing error rates to maximum possible estimate.</span>
<span class="co">#&gt; selfConsist step 1 .....</span>
<span class="co">#&gt;    selfConsist step 2</span>
<span class="co">#&gt;    selfConsist step 3</span>
<span class="co">#&gt;    selfConsist step 4</span>
<span class="co">#&gt; Convergence after  4  rounds.</span>

<span class="va">MA.alt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mergeMulti.html">mergeMulti</a></span><span class="op">(</span><span class="va">MA.alt</span>, justConcatenate<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>

<span class="va">MA.alt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeSequenceTableMulti.html">makeSequenceTableMulti</a></span><span class="op">(</span><span class="va">MA.alt</span><span class="op">)</span>

<span class="va">MA.alt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/removeChimeraMulti.html">removeChimeraMulti</a></span><span class="op">(</span><span class="va">MA.alt</span>, mc.cores<span class="op">=</span><span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>So we can make a comparison between the two different versions:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">together</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MultiAmplicon-class.html">getSequencesFromTable</a></span><span class="op">(</span><span class="va">MA6</span><span class="op">)</span>
<span class="va">alt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MultiAmplicon-class.html">getSequencesFromTable</a></span><span class="op">(</span><span class="va">MA.alt</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span><span class="op">(</span><span class="va">together</span><span class="op">)</span>, <span class="kw">function</span> <span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">alt</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">%in%</span><span class="va">together</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; FALSE  TRUE </span>
<span class="co">#&gt;     6   219 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; FALSE  TRUE </span>
<span class="co">#&gt;    11   153 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[3]]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; FALSE  TRUE </span>
<span class="co">#&gt;     2   157 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[4]]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; FALSE  TRUE </span>
<span class="co">#&gt;     2   116 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[5]]</span>
<span class="co">#&gt; &lt; table of extent 0 &gt;</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[6]]</span>
<span class="co">#&gt; &lt; table of extent 0 &gt;</span></code></pre></div>
<p>So in this case we got nearly the same output: 215 of 8 (for the first amplicon) and even higher proportions of identical speaks for error rates not differing between amplicons (and for the robustness of the ASV inference of the amazing dada2). Which one you should use depends on your data: if each amplicon has enough reads you can estimate seperately and potentially get a better estimate of errors for amplicons with different errors. If on the other hand some amplicons have very few reads, it might make sense to pool them to improve error estimates with larger size for the training set.</p>
</div>
<div id="using-different-parameters-for-each-amplicon" class="section level2">
<h2 class="hasAnchor">
<a href="#using-different-parameters-for-each-amplicon" class="anchor"></a>Using different parameters for each amplicon</h2>
<p>In some real world applications it might be desirable to have more control over parameters used in the pipeline for different amplicons.</p>
<p>We use ellipsis (aka “…”, dots, dot-dot-dot or three-dots) to pass parameters on to the underlying dada2 functions. The MultiAmplicon package allows all those arguments to be vectors with values varying for different amplicons.</p>
<p>Let’s suppose for this we e.g. want to merge a subset of our amplicons, while we concatenate others (with Ns between forward and reverser reads).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">MA.mixed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mergeMulti.html">mergeMulti</a></span><span class="op">(</span><span class="va">MA3</span>, justConcatenate<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">FALSE</span>, <span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 6 samples for amplicon Amp1F.Amp1R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 5 samples for amplicon Amp2F.Amp2R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=FALSE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 2 samples for amplicon Amp3F.Amp3R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=FALSE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 5 samples for amplicon Amp4F.Amp4R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 0 samples for amplicon Amp5F.Amp5R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 0 samples for amplicon Amp6F.Amp6R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span></code></pre></div>
<p>The above example would make sense if (only) amplicons Amp1F.Amp1R and Amp4F.Amp4R (1st and 4th in our object) had too little overlap between forward and reverse reads to be merged and would better be concatenated. Btw: Here this makes not really sense as to much sequences for the 2nd and 3rd amplicon don’t merge for this dataset.</p>
<p>Coming back to our example on dada sequence inference, we could also mix err estimates to use errors either seperately or accross amplicons. Always set the references to both pre-computed error parameters <code>NULL</code> when using self consistent estimation within one amplicon.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">MA.mixed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dadaMulti.html">dadaMulti</a></span><span class="op">(</span><span class="va">MA3</span>,
                      Ferr<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">NULL</span>, <span class="va">errF</span>,  <span class="cn">NULL</span>, <span class="va">errF</span><span class="op">)</span>,
                      Rerr<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">NULL</span>, <span class="va">errF</span>,  <span class="cn">NULL</span>, <span class="va">errF</span><span class="op">)</span>,
               selfConsist<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">TRUE</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; amplicon Amp1F.Amp1R: dada estimation of sequence variants from 6 of 8 possible sample files</span>
<span class="co">#&gt; calling dada with selfConsist=TRUE parameters</span>
<span class="co">#&gt; selfConsist step 1 ......</span>
<span class="co">#&gt;    selfConsist step 2</span>
<span class="co">#&gt;    selfConsist step 3</span>
<span class="co">#&gt;    selfConsist step 4</span>
<span class="co">#&gt; Convergence after  4  rounds.</span>
<span class="co">#&gt; selfConsist step 1 ......</span>
<span class="co">#&gt;    selfConsist step 2</span>
<span class="co">#&gt;    selfConsist step 3</span>
<span class="co">#&gt;    selfConsist step 4</span>
<span class="co">#&gt; Convergence after  4  rounds.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; amplicon Amp2F.Amp2R: dada estimation of sequence variants from 5 of 8 possible sample files</span>
<span class="co">#&gt; calling dada with selfConsist=FALSE parameters</span>
<span class="co">#&gt; Sample 1 - 2949 reads in 1142 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 2 reads in 2 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 20391 reads in 4136 unique sequences.</span>
<span class="co">#&gt; Sample 4 - 20391 reads in 4136 unique sequences.</span>
<span class="co">#&gt; Sample 5 - 2690 reads in 968 unique sequences.</span>
<span class="co">#&gt; Sample 1 - 2949 reads in 1290 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 2 reads in 2 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 20391 reads in 5719 unique sequences.</span>
<span class="co">#&gt; Sample 4 - 20391 reads in 5719 unique sequences.</span>
<span class="co">#&gt; Sample 5 - 2690 reads in 1100 unique sequences.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; amplicon Amp3F.Amp3R: dada estimation of sequence variants from 2 of 8 possible sample files</span>
<span class="co">#&gt; calling dada with selfConsist=TRUE parameters</span>
<span class="co">#&gt; selfConsist step 1 ..</span>
<span class="co">#&gt;    selfConsist step 2</span>
<span class="co">#&gt;    selfConsist step 3</span>
<span class="co">#&gt;    selfConsist step 4</span>
<span class="co">#&gt; Convergence after  4  rounds.</span>
<span class="co">#&gt; selfConsist step 1 ..</span>
<span class="co">#&gt;    selfConsist step 2</span>
<span class="co">#&gt;    selfConsist step 3</span>
<span class="co">#&gt;    selfConsist step 4</span>
<span class="co">#&gt; Convergence after  4  rounds.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; amplicon Amp4F.Amp4R: dada estimation of sequence variants from 5 of 8 possible sample files</span>
<span class="co">#&gt; calling dada with selfConsist=FALSE parameters</span>
<span class="co">#&gt; Sample 1 - 1087 reads in 394 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 7 reads in 6 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 2076 reads in 959 unique sequences.</span>
<span class="co">#&gt; Sample 4 - 2076 reads in 959 unique sequences.</span>
<span class="co">#&gt; Sample 5 - 428 reads in 166 unique sequences.</span>
<span class="co">#&gt; Sample 1 - 1087 reads in 433 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 7 reads in 5 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 2076 reads in 962 unique sequences.</span>
<span class="co">#&gt; Sample 4 - 2076 reads in 962 unique sequences.</span>
<span class="co">#&gt; Sample 5 - 428 reads in 183 unique sequences.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; amplicon Amp5F.Amp5R: dada estimation of sequence variants from 0 of 8 possible sample files</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; skipping empty amplicon</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; amplicon Amp6F.Amp6R: dada estimation of sequence variants from 0 of 8 possible sample files</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; skipping empty amplicon</span></code></pre></div>
</div>
<div id="taxnomical-annotation" class="section level2">
<h2 class="hasAnchor">
<a href="#taxnomical-annotation" class="anchor"></a>Taxnomical annotation</h2>
<p>Taxonimical annotation is difficult as for 18S markers reference databases are not as mature as for 16S. As we use Multiamplicon in our own research it includes for now only a 18S annotation via blast. This only works on</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">MA7</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/blastTaxAnnot.html">blastTaxAnnot</a></span><span class="op">(</span><span class="va">MA6</span>, db<span class="op">=</span><span class="st">"/SAN/db/blastdb/nt/nt"</span>,
                     negative_gilist <span class="op">=</span> <span class="st">"/SAN/db/blastdb/uncultured_gilist.txt"</span>,
                     taxonSQL<span class="op">=</span><span class="st">"/SAN/db/taxonomy/taxonomizr.sql"</span>, num_threads<span class="op">=</span><span class="fl">20</span><span class="op">)</span>
<span class="co">#&gt; wrote file /tmp/Rtmp15p2UW/filecc847a6f5bfc.fasta storing input sequences for blast</span>
<span class="co">#&gt; using database /SAN/db/blastdb/nt/nt resulting in as blast database full path</span>
<span class="co">#&gt; STARTED running blast with command:</span>
<span class="co">#&gt; blastn -negative_gilist /SAN/db/blastdb/uncultured_gilist.txt -query /tmp/Rtmp15p2UW/filecc847a6f5bfc.fasta -db /SAN/db/blastdb/nt/nt -evalue 1e-15 -num_threads 20 -max_hsps 1 -max_target_seqs 25 -out /tmp/Rtmp15p2UW/filecc8413791925.blt -outfmt '10 qaccver saccver pident length mismatch gapopen qstart qend sstart send evalue bitscore staxid'</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; FINISHED running blast</span></code></pre></div>
</div>
<div id="handing-over-to-phyloseq" class="section level2">
<h2 class="hasAnchor">
<a href="#handing-over-to-phyloseq" class="anchor"></a>Handing over to phyloseq</h2>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">PHY</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/toPhyloseq.html">toPhyloseq</a></span><span class="op">(</span><span class="va">MA7</span>, sample<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">MA7</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; extracting 8 requested samples</span>
<span class="co">#&gt; filling zeros for amplicon missing 3 samples</span>
<span class="co">#&gt; filling zeros for amplicon missing 4 samples</span>
<span class="co">#&gt; filling zeros for amplicon missing 6 samples</span>
<span class="co">#&gt; filling zeros for amplicon missing 4 samples</span>
<span class="co">#&gt; filling zeros for amplicon missing 8 samples </span>
<span class="co">#&gt; filling zeros for amplicon missing 8 samples</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Emanuel Heitlinger.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
