<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>MultiAmplicon - a real world example • MultiAmplicon</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="MultiAmplicon - a real world example">
<meta property="og:description" content="MultiAmplicon">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MultiAmplicon</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/MultiAmplicon-real-world-example.html">MultiAmplicon - a real world example</a>
    </li>
    <li>
      <a href="../articles/MultiAmplicon-small-example.html">MultiAmplicon - a small example</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/derele/MultiAmplicon/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>MultiAmplicon - a real world example</h1>
                        <h4 class="author">Emanuel Heitlinger</h4>
            
            <h4 class="date">2021-01-20</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/derele/MultiAmplicon/blob/master/vignettes/MultiAmplicon-real-world-example.Rmd"><code>vignettes/MultiAmplicon-real-world-example.Rmd</code></a></small>
      <div class="hidden name"><code>MultiAmplicon-real-world-example.Rmd</code></div>

    </div>

    
    
<div id="data-preparation-download-and-filtering" class="section level2">
<h2 class="hasAnchor">
<a href="#data-preparation-download-and-filtering" class="anchor"></a>Data preparation (download and filtering)</h2>
<p>We first download the dataset consisting of 192 fastq file-pairs from 18S and 16S amplicon sequencing of feces and intestinal contents of carnivores (hyenas and wolves). The help files <a href="https://derele.github.io/MultiAmplicon/reference/carnivoreSeqRuns.html"><code>carnivoreSeqRuns</code></a> shows how this data has been downloaded from <a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA386767">NCBI-SRA</a>.</p>
<p>After download we can run our standard workflow. But as usual with real data we have to filter the input files first. This particular sequencing was of quite low quality, so we have to trim and screen quite harshly.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://derele.github.io/MultiAmplicon/">MultiAmplicon</a></span><span class="op">)</span>
<span class="co">## And we'll also use some dada2 functions directly</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://benjjneb.github.io/dada2/">dada2</a></span><span class="op">)</span>
<span class="va">path</span> <span class="op">&lt;-</span> <span class="st">"download_sra"</span> <span class="co">## change according to where you downloaded</span>

<span class="va">fastqFiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">path</span>, pattern<span class="op">=</span><span class="st">".fastq.gz$"</span>, full.names<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>

<span class="va">fastqF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"_1.fastq.gz"</span>, <span class="va">fastqFiles</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">fastqR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"_2.fastq.gz"</span>, <span class="va">fastqFiles</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"_1.fastq\\.gz"</span>, <span class="st">"\\1"</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">fastqF</span><span class="op">)</span><span class="op">)</span>

<span class="va">filt_path</span> <span class="op">&lt;-</span> <span class="st">"filtered_sra"</span>
<span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/utils/filetest.html">file_test</a></span><span class="op">(</span><span class="st">"-d"</span>, <span class="va">filt_path</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">filt_path</span><span class="op">)</span>

<span class="va">filtFs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">filt_path</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">samples</span>, <span class="st">"_F_filt.fastq.gz"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">filtFs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">samples</span>
<span class="va">filtRs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">filt_path</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">samples</span>, <span class="st">"_R_filt.fastq.gz"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">filtRs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">samples</span>

<span class="co">## some files will be filtered out completely, therefore allowing 50</span>
<span class="co">## files less present and still don't redo filtering</span>
<span class="va">redo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span><span class="va">fastqF</span><span class="op">)</span><span class="op">)</span> <span class="op">-</span>
    <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span><span class="va">filtFs</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">50</span>

<span class="kw">if</span><span class="op">(</span><span class="va">redo</span><span class="op">)</span><span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span><span class="op">(</span><span class="va">fastqF</span><span class="op">)</span>,  <span class="kw">function</span> <span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span>
        <span class="fu"><a href="https://rdrr.io/pkg/dada2/man/filterAndTrim.html">filterAndTrim</a></span><span class="op">(</span><span class="va">fastqF</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="va">filtFs</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="va">fastqR</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="va">filtRs</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,
                      truncLen<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">170</span>,<span class="fl">170</span><span class="op">)</span>, minLen<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">170</span>,<span class="fl">170</span><span class="op">)</span>, 
                      maxN<span class="op">=</span><span class="fl">0</span>, maxEE<span class="op">=</span><span class="fl">2</span>, truncQ<span class="op">=</span><span class="fl">2</span>, 
                      compress<span class="op">=</span><span class="cn">TRUE</span>, verbose<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
    <span class="op">}</span><span class="op">)</span>
<span class="op">}</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">filtFs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">filtRs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">samples</span>

<span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PairedReadFileSet-class.html">PairedReadFileSet</a></span><span class="op">(</span><span class="va">filtFs</span>, <span class="va">filtRs</span><span class="op">)</span>
<span class="co">#&gt; Warning in validityMethod(object): </span>
<span class="co">#&gt; file filtered_sra/SRR5569127_F_filt.fastq.gz does not exist on your system</span>
<span class="co">#&gt; file filtered_sra/SRR5569127_R_filt.fastq.gz does not exist on your system</span></code></pre></div>
</div>
<div id="data-preparation-primer-set" class="section level2">
<h2 class="hasAnchor">
<a href="#data-preparation-primer-set" class="anchor"></a>Data preparation (primer set)</h2>
<p>Now we have our sequencing read input data, only the primers are still missing. The primer set for the above is included in the MultiAmplicon package. The data has been generated using a microfluidics PCR system, if you want to know more about this you can consult <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5472691/">Heitlinger et al. (2017)</a> or <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5269671/">Lesniak et al. (2017)</a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">primer.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"real_world_primers.csv"</span>,
                           package <span class="op">=</span> <span class="st">"MultiAmplicon"</span><span class="op">)</span>

<span class="va">ptable</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="va">primer.file</span>, sep<span class="op">=</span><span class="st">","</span>, header<span class="op">=</span><span class="cn">TRUE</span>, stringsAsFactors<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>

<span class="va">primerF</span> <span class="op">&lt;-</span> <span class="va">ptable</span><span class="op">[</span>, <span class="st">"TS.SequenceF"</span><span class="op">]</span>
<span class="va">primerR</span> <span class="op">&lt;-</span> <span class="va">ptable</span><span class="op">[</span>, <span class="st">"TS.SequenceR"</span><span class="op">]</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">primerF</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">ptable</span><span class="op">[</span>, <span class="st">"corrected.NameF"</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">primerR</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">ptable</span><span class="op">[</span>, <span class="st">"corrected.NameR"</span><span class="op">]</span><span class="op">)</span>

<span class="va">primers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PrimerPairsSet-class.html">PrimerPairsSet</a></span><span class="op">(</span><span class="va">primerF</span>, <span class="va">primerR</span><span class="op">)</span>

<span class="va">MA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MultiAmplicon-class.html">MultiAmplicon</a></span><span class="op">(</span><span class="va">primers</span>, <span class="va">files</span><span class="op">)</span></code></pre></div>
</div>
<div id="running-the-multiamplicon-pipeline" class="section level2">
<h2 class="hasAnchor">
<a href="#running-the-multiamplicon-pipeline" class="anchor"></a>Running the MultiAmplicon pipeline</h2>
<p>We start by sorting our amplicons by primer sequences cutting off the latter from sequencing reads. The directory for sorted amplicons must be empty before that.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">filedir</span> <span class="op">&lt;-</span> <span class="st">"stratified_files"</span>
<span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.exists</a></span><span class="op">(</span><span class="va">filedir</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/unlink.html">unlink</a></span><span class="op">(</span><span class="va">filedir</span>, recursive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="va">MA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sortAmplicons.html">sortAmplicons</a></span><span class="op">(</span><span class="va">MA</span>, n<span class="op">=</span><span class="fl">1e+05</span>, filedir<span class="op">=</span><span class="va">filedir</span><span class="op">)</span></code></pre></div>
<p>The function is “streaming” fastq files, so that potentially large files can be handled on machines with modest memory. The parameter “n” controls how big the streamed chunks are. Increasing this value leads to faster processing and higher memory usage.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">errF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/dada2/man/learnErrors.html">learnErrors</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="../reference/MultiAmplicon-class.html">getStratifiedFilesF</a></span><span class="op">(</span><span class="va">MA</span><span class="op">)</span><span class="op">)</span>, nbase<span class="op">=</span><span class="fl">1e8</span>,
                    verbose<span class="op">=</span><span class="fl">0</span><span class="op">)</span>
<span class="va">errR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/dada2/man/learnErrors.html">learnErrors</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="../reference/MultiAmplicon-class.html">getStratifiedFilesR</a></span><span class="op">(</span><span class="va">MA</span><span class="op">)</span><span class="op">)</span>, nbase<span class="op">=</span><span class="fl">1e8</span>,
                    verbose<span class="op">=</span><span class="fl">0</span><span class="op">)</span>

<span class="va">MA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dadaMulti.html">dadaMulti</a></span><span class="op">(</span><span class="va">MA</span>, Ferr<span class="op">=</span><span class="va">errF</span>, Rerr<span class="op">=</span><span class="va">errR</span>,  pool<span class="op">=</span><span class="cn">FALSE</span>,
                verbose<span class="op">=</span><span class="fl">0</span><span class="op">)</span></code></pre></div>
<p>The sequences were quite low quality and we shortened them massively. Therefore we have to assume that most sequences won’t have enough overlap to merge. Now we figure out for which amplicons this is the case.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">MA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mergeMulti.html">mergeMulti</a></span><span class="op">(</span><span class="va">MA</span><span class="op">)</span>

<span class="va">propMerged</span> <span class="op">&lt;-</span> <span class="fu">MultiAmplicon</span><span class="fu">::</span><span class="fu"><a href="../reference/calcPropMerged.html">calcPropMerged</a></span><span class="op">(</span><span class="va">MA</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">propMerged</span><span class="op">)</span>
<span class="co">#&gt;      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. </span>
<span class="co">#&gt; 0.0000000 0.0002201 0.0013316 0.0416265 0.0063945 0.8817513</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">propMerged</span><span class="op">&lt;</span><span class="fl">0.8</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; FALSE  TRUE </span>
<span class="co">#&gt;     1    47</span></code></pre></div>
<p>So we just found out that for this dataset only one amplicon could be merged retaining over 80% of the sequence. Lets run the merge again, concatenating all amplicons with less than 80% merged by setting <code>justConcatenate</code> to <code>TRUE</code> for those.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">MA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mergeMulti.html">mergeMulti</a></span><span class="op">(</span><span class="va">MA</span>, justConcatenate<span class="op">=</span><span class="va">propMerged</span><span class="op">&lt;</span><span class="fl">0.8</span><span class="op">)</span>
<span class="va">MA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeSequenceTableMulti.html">makeSequenceTableMulti</a></span><span class="op">(</span><span class="va">MA</span><span class="op">)</span>
<span class="va">MA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/removeChimeraMulti.html">removeChimeraMulti</a></span><span class="op">(</span><span class="va">MA</span><span class="op">)</span></code></pre></div>
</div>
<div id="check-what-happend-within-the-pipeline" class="section level2">
<h2 class="hasAnchor">
<a href="#check-what-happend-within-the-pipeline" class="anchor"></a>Check what happend within the pipeline</h2>
<p>When we are done with our processing, we usually want to know what happened to our raw data (sequencing reads) along the MultiAmplicon/dada2 pipeline.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tracking</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getPipelineSummary.html">getPipelineSummary</a></span><span class="op">(</span><span class="va">MA</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotPipelineSummary.html">plotPipelineSummary</a></span><span class="op">(</span><span class="va">tracking</span><span class="op">)</span></code></pre></div>
<p><img src="MultiAmplicon-real-world-example_files/figure-html/tracking-1.png" width="700"></p>
<p>The pipeline summary tells us how many samples, how many unique sequences and how many sequence read pairs were recovered at different steps through the pipeline. It gives this information for the different amplicons.</p>
<p>The output of the pipeline summary can be modified with standard ggplot syntax.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="fu"><a href="../reference/plotPipelineSummary.html">plotPipelineSummary</a></span><span class="op">(</span><span class="va">tracking</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="MultiAmplicon-real-world-example_files/figure-html/plottrack-1.png" width="700"></p>
</div>
<div id="hand-over-to-phyloseq" class="section level2">
<h2 class="hasAnchor">
<a href="#hand-over-to-phyloseq" class="anchor"></a>Hand over to phyloseq</h2>
<p>We haven’t genereated a taxonomical annotation for this ASVs yet. oing so within MulitAmplicon is currently only supported on Unix systems (see <code>blastTaxAnnot</code>). You will need to generat a taxonomical annotation in one way or the other at some point, but here we demonstrate ho to do an export to a phyloseq object first.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217">phyloseq</a></span><span class="op">)</span>
<span class="va">PH</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/toPhyloseq.html">toPhyloseq</a></span><span class="op">(</span><span class="va">MA</span>, samples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">MA</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>We here produced a combined phyloseq object including ASVs from all different amplicons.</p>
</div>
<div id="comparing-sample-data" class="section level2">
<h2 class="hasAnchor">
<a href="#comparing-sample-data" class="anchor"></a>Comparing sample data</h2>
<p>We now can use phyloseq for one of the many supported analyses.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/plot_richness.html">plot_richness</a></span><span class="op">(</span><span class="va">PH</span>, measures<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Observed"</span>, <span class="st">"Chao1"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="MultiAmplicon-real-world-example_files/figure-html/phydiversity-1.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Emanuel Heitlinger.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
