<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>MultiAmplicon - a real world example • MultiAmplicon</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">MultiAmplicon</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/MultiAmplicon-real-world-example.html">MultiAmplicon - a real world example</a>
    </li>
    <li>
      <a href="../articles/MultiAmplicon-small-example.html">MultiAmplicon - a small example</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>MultiAmplicon - a real world example</h1>
                        <h4 class="author">Emanuel Heitlinger</h4>
            
            <h4 class="date">2018-04-17</h4>
          </div>

    
    
<div class="contents">
<div id="data-preparation-download-and-filtering" class="section level2">
<h2 class="hasAnchor">
<a href="#data-preparation-download-and-filtering" class="anchor"></a>Data preparation (download and filtering)</h2>
<p>We first <a href="https://svalbard.biologie.hu-berlin.de:443/d/c291b870178f4abc8c59/">download the data</a> set from consisting of 192 fastq file-pairs from 18S and 16S amplicon sequencing of feces and intestinal contents of carnivores (hyenas and wolves). Use the botton on the upper right of the page to download a zipped folder and unzip it after downloading. The help file for the <a href="https://derele.github.io/MultiAmplicon/reference/carnivoreSeqRuns.html"><code>carnivoreSeqRuns</code></a> (containing sample data included in the package), shows how this data has been downloaded from <a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA386767">NCBI-SRA</a>.</p>
<p>Now we can run our standard workflow. But as usual with real data we have to filter the input files first. This particular sequencing was of quite low quality, so we have to trim and screen quite harshly.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(MultiAmplicon)

path &lt;-<span class="st"> "~/download_sra"</span> ## change according to where you downloaded

fastqFiles &lt;-<span class="st"> </span><span class="kw">list.files</span>(path, <span class="dt">pattern=</span><span class="st">".fastq.gz$"</span>, <span class="dt">full.names=</span><span class="ot">TRUE</span>)

fastqF &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="st">"_1.fastq.gz"</span>, fastqFiles, <span class="dt">value =</span> <span class="ot">TRUE</span>)
fastqR &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="st">"_2.fastq.gz"</span>, fastqFiles, <span class="dt">value =</span> <span class="ot">TRUE</span>)

samples &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">"_1.fastq</span><span class="ch">\\</span><span class="st">.gz"</span>, <span class="st">"</span><span class="ch">\\</span><span class="st">1"</span>, <span class="kw">basename</span>(fastqF))

filt_path &lt;-<span class="st"> "~/filtered_sra/"</span>
if(!<span class="kw">file_test</span>(<span class="st">"-d"</span>, filt_path)) <span class="kw">dir.create</span>(filt_path)

filtFs &lt;-<span class="st"> </span><span class="kw">file.path</span>(filt_path, <span class="kw">paste0</span>(samples, <span class="st">"_F_filt.fastq.gz"</span>))
<span class="kw">names</span>(filtFs) &lt;-<span class="st"> </span>samples
filtRs &lt;-<span class="st"> </span><span class="kw">file.path</span>(filt_path, <span class="kw">paste0</span>(samples, <span class="st">"_R_filt.fastq.gz"</span>))
<span class="kw">names</span>(filtRs) &lt;-<span class="st"> </span>samples

## some files will be filtered out completely, therefore allowing 50
## files less present and still don't redo filtering
if(<span class="kw">sum</span>(<span class="kw">file.exists</span>(fastqF)) -
<span class="st">   </span><span class="kw">sum</span>(<span class="kw">file.exists</span>(filtFs)) &gt;<span class="st"> </span><span class="dv">50</span>){
    <span class="kw">lapply</span>(<span class="kw">seq_along</span>(fastqF),  function (i) {
        <span class="kw">filterAndTrim</span>(fastqF[i], filtFs[i], fastqR[i], filtRs[i],
                      <span class="dt">truncLen=</span><span class="kw">c</span>(<span class="dv">170</span>,<span class="dv">170</span>), <span class="dt">minLen=</span><span class="kw">c</span>(<span class="dv">170</span>,<span class="dv">170</span>), 
                      <span class="dt">maxN=</span><span class="dv">0</span>, <span class="dt">maxEE=</span><span class="dv">2</span>, <span class="dt">truncQ=</span><span class="dv">2</span>, 
                      <span class="dt">compress=</span><span class="ot">TRUE</span>, <span class="dt">verbose=</span><span class="ot">TRUE</span>)
    })
}

<span class="kw">names</span>(filtFs) &lt;-<span class="st"> </span><span class="kw">names</span>(filtRs) &lt;-<span class="st"> </span>samples

files &lt;-<span class="st"> </span><span class="kw"><a href="../reference/PairedReadFileSet-class.html">PairedReadFileSet</a></span>(filtFs, filtRs)
<span class="co">#&gt; Warning in validityMethod(object): </span>
<span class="co">#&gt; file ~/filtered_sra//SRR5569127_F_filt.fastq.gz does not exist on your system</span>
<span class="co">#&gt; file ~/filtered_sra//SRR5569127_R_filt.fastq.gz does not exist on your system</span></code></pre></div>
</div>
<div id="data-preparation-primer-set" class="section level2">
<h2 class="hasAnchor">
<a href="#data-preparation-primer-set" class="anchor"></a>Data preparation (primer set)</h2>
<p>Now we have our sequencing read input data, only the primers are still missing. The primer set for the above is included in the MultiAmplicon package. The data has been generated using a microfluidics PCR system, if you want to know more about this you can consult <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5472691/">Heitlinger et al. (2017)</a> or <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5269671/">Lesniak et al. (2017)</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">primer.file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"real_world_primers.csv"</span>,
                           <span class="dt">package =</span> <span class="st">"MultiAmplicon"</span>)

ptable &lt;-<span class="st"> </span><span class="kw">read.csv</span>(primer.file, <span class="dt">sep=</span><span class="st">","</span>, <span class="dt">header=</span><span class="ot">TRUE</span>, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)

primerF &lt;-<span class="st"> </span>ptable[, <span class="st">"TS.SequenceF"</span>]
primerR &lt;-<span class="st"> </span>ptable[, <span class="st">"TS.SequenceR"</span>]

<span class="kw">names</span>(primerF) &lt;-<span class="st"> </span><span class="kw">as.character</span>(ptable[, <span class="st">"corrected.NameF"</span>])
<span class="kw">names</span>(primerR) &lt;-<span class="st"> </span><span class="kw">as.character</span>(ptable[, <span class="st">"corrected.NameR"</span>])

primers &lt;-<span class="st"> </span><span class="kw"><a href="../reference/PrimerPairsSet-class.html">PrimerPairsSet</a></span>(primerF, primerR)

MA &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MultiAmplicon-class.html">MultiAmplicon</a></span>(primers, files)</code></pre></div>
</div>
<div id="running-the-multiamplicon-pipeline" class="section level2">
<h2 class="hasAnchor">
<a href="#running-the-multiamplicon-pipeline" class="anchor"></a>Running the MultiAmplicon pipeline</h2>
<p>We start by sorting our amplicons by primer sequences cutting off the latter from sequencing reads.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MA &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sortAmplicons.html">sortAmplicons</a></span>(MA)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">errF &lt;-<span class="st"> </span><span class="kw">learnErrors</span>(<span class="kw">unlist</span>(<span class="kw"><a href="../reference/MultiAmplicon-class.html">getStratifiedFilesF</a></span>(MA)), <span class="dt">nread=</span><span class="fl">1e6</span>,
                    <span class="dt">verbose=</span><span class="dv">0</span>)
errR &lt;-<span class="st"> </span><span class="kw">learnErrors</span>(<span class="kw">unlist</span>(<span class="kw"><a href="../reference/MultiAmplicon-class.html">getStratifiedFilesR</a></span>(MA)), <span class="dt">nread=</span><span class="fl">1e6</span>,
                    <span class="dt">verbose=</span><span class="dv">0</span>)

MA &lt;-<span class="st"> </span><span class="kw"><a href="../reference/derepMulti.html">derepMulti</a></span>(MA, <span class="dt">mc.cores=</span><span class="dv">1</span>) 
MA &lt;-<span class="st"> </span><span class="kw"><a href="../reference/dadaMulti.html">dadaMulti</a></span>(MA, <span class="dt">Ferr=</span>errF, <span class="dt">Rerr=</span>errR,  <span class="dt">pool=</span><span class="ot">FALSE</span>,
                <span class="dt">verbose=</span><span class="dv">0</span>)</code></pre></div>
<p>The sequences were quite low quality and we shortened them massively. Therefore we have to assume that most sequences won’t have enough overlap to merge. Now we figure out for which amplicons this is the case.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MA &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mergeMulti.html">mergeMulti</a></span>(MA)

propMerged &lt;-<span class="st"> </span>MultiAmplicon::<span class="kw"><a href="../reference/calcPropMerged.html">calcPropMerged</a></span>(MA)

<span class="kw">summary</span>(propMerged)
<span class="co">#&gt;      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. </span>
<span class="co">#&gt; 0.0000000 0.0001524 0.0006254 0.0367071 0.0037608 0.8828855</span>
<span class="kw">table</span>(propMerged&lt;<span class="fl">0.8</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt; FALSE  TRUE </span>
<span class="co">#&gt;     1    47</span></code></pre></div>
<p>So we just found out that for this dataset only one amplicon could be merged retaining over 80% of the sequence. Lets run the merge again, concatenating all amplicons with less than 80% merged by setting <code>justConcatenate</code> to <code>TRUE</code> for those.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MA &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mergeMulti.html">mergeMulti</a></span>(MA, <span class="dt">justConcatenate=</span>propMerged&lt;<span class="fl">0.8</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 179 samples for amplicon ACM_008.ADM330_CR</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 180 samples for amplicon MarkN.Amann</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 169 samples for amplicon Medin.Amann</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 185 samples for amplicon MarkN.Euk360_CR</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 180 samples for amplicon proti1702.EukB_MH</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 188 samples for amplicon Nem_0425_6CR.EukB_MH</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 173 samples for amplicon Hadz_1200CR.EukB_MH</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 187 samples for amplicon Nem_0425_6CR.EukB_MV</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 184 samples for amplicon proti1702.EukB_MV</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 167 samples for amplicon Hadz_1200CR.EukB_MV</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 143 samples for amplicon Medin.Gen_600_CR</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 171 samples for amplicon Hadz_566.Hadz_1200</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 184 samples for amplicon Gen_600.Hadz_1200</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 165 samples for amplicon AmannCR.Hadz_1200</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 181 samples for amplicon Mach1.Hadz_1200</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 191 samples for amplicon ACM_008.Klin0341_CR</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 190 samples for amplicon ADM330.Klin0785_CR</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 188 samples for amplicon Klin0341.Klin0785_CR</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 157 samples for amplicon MarkN.Mach1CR</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 166 samples for amplicon Gen_600.Mach2Gen</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 175 samples for amplicon Hadz_566.Mach2Gen</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 164 samples for amplicon AmannCR.Mach2Gen</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 179 samples for amplicon Mach1.Mach2Gen</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 179 samples for amplicon MarkN.Mark</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 181 samples for amplicon Mach1.Nem_0425_2</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 174 samples for amplicon Hadz_566.Nem_0425_4R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 183 samples for amplicon Gen_600.Nem_0425_4R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 179 samples for amplicon Mach1.Nem_0425_4R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 172 samples for amplicon Nem_0425_4F.Nem_0425_4R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 188 samples for amplicon wang1141.Nem_0425_6</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=FALSE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 145 samples for amplicon proti440.proti1580_R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 165 samples for amplicon proti643.proti1580_R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 147 samples for amplicon proti482.proti1580_R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 187 samples for amplicon proti15.proti440_R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 188 samples for amplicon MarkN.proti440_R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 159 samples for amplicon proti15.proti482_R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 138 samples for amplicon MarkN.proti628_R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 102 samples for amplicon proti89.proti628_R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 125 samples for amplicon proti15.proti628_R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 174 samples for amplicon proti15.proti643_R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 182 samples for amplicon proti89.proti643_R</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 171 samples for amplicon proti440.wang1141CR</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 171 samples for amplicon proti643.wang1141CR</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 171 samples for amplicon proti628.wang1141CR</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 187 samples for amplicon proti1580.wang1624CR6L</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 184 samples for amplicon Hadz_1200CR.wang1624CR6L</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 189 samples for amplicon proti1702.wang1624CR6L</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from 165 samples for amplicon Hadz_1200CR.wang1624CR6S</span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
MA &lt;-<span class="st"> </span><span class="kw"><a href="../reference/makeSequenceTableMulti.html">makeSequenceTableMulti</a></span>(MA)
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; The sequences being tabled vary in length.</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
<span class="co">#&gt; calling makeSequenceTable with default parameters</span>
MA &lt;-<span class="st"> </span><span class="kw"><a href="../reference/removeChimeraMulti.html">removeChimeraMulti</a></span>(MA, <span class="dt">mc.cores=</span><span class="dv">1</span>)
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
<span class="co">#&gt; calling removeBimeraDenovo with default parameters</span>
<span class="co">#&gt; As of the 1.4 release, the default method changed to consensus (from pooled).</span>
MA &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fillSampleTables.html">fillSampleTables</a></span>(MA)</code></pre></div>
</div>
<div id="hand-over-to-phyloseq" class="section level2">
<h2 class="hasAnchor">
<a href="#hand-over-to-phyloseq" class="anchor"></a>Hand over to phyloseq</h2>
</div>
<div id="comparing-sample-data" class="section level2">
<h2 class="hasAnchor">
<a href="#comparing-sample-data" class="anchor"></a>Comparing sample data</h2>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#data-preparation-download-and-filtering">Data preparation (download and filtering)</a></li>
      <li><a href="#data-preparation-primer-set">Data preparation (primer set)</a></li>
      <li><a href="#running-the-multiamplicon-pipeline">Running the MultiAmplicon pipeline</a></li>
      <li><a href="#hand-over-to-phyloseq">Hand over to phyloseq</a></li>
      <li><a href="#comparing-sample-data">Comparing sample data</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Emanuel Heitlinger.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
